### ISSUE TITLE
Fish 3.0 does not run under Cygwin

### ISSUE THREAD

**Issue by zanchey** [fish-shell/fish-shell#5426]
The latest build from the Integration_3.0.0 (d6e315d25db5c2a1) and master (50fbc36b) fails to launch, and fails the tests at the Ctrl-C cancellation by hanging indefinitely, under Cygwin.

Running `fish -d5` leaves me with an infinitely-looping output that finishes up like this:

```
<4> fish: proc::read_try('status current-command')
<4> fish: io_buffer_t::read: blocking read on fd 5
<4> fish: Created pipe using fds 5 and 6
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/fish_title.fish
<4> fish: Created pipe using fds 7 and 8
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/fish_title.fish
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/__fish_pwd.fish
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/fish_title.fish
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/__fish_pwd.fish
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/__fish_pwd.fish
<4> fish: Created pipe using fds 9 and 10
<5> fish: intern /home/zanchey/src/fish-shell/share/functions/__fish_pwd.fish
<5> fish: path_get_path( 'uname' )
<4> fish: env_export_arr() recalc
<4> fish: Fork #1, pid 5984: spawn external command '/usr/bin/uname' from '/home/zanchey/src/fish-shell/share/functions/__fish_pwd.fish'
<3> fish: Created job 5 from command 'uname' with pgrp 5384
<4> fish: Start job 5, gid 5384 (uname), UNCOMPLETED, NON-INTERACTIVE
<4> fish: select_try on fd 9
<4> fish: select_try hit timeout
<4> fish: select_try: no fds returned valid data within the timeout
<4> fish: select_try on fd 9
<4> fish: proc::read_try('uname')
<5> fish: Skipping wait on incomplete job 4 (source ...)
<5> fish: Skipping wait on incomplete job 1 (fish_title ...)
<4> fish: select_try on fd 9
<4> fish: select_try hit timeout
<4> fish: select_try: no fds returned valid data within the timeout
<5> fish: Skipping wait on incomplete job 4 (source ...)
<5> fish: Skipping wait on incomplete job 1 (fish_title ...)
```
(repeats from `select_try on fd 9` forever)

Getting builds on Cygwin involves an astonishing amount of yak-shaving for me (recent searches include "freebsd nfs tuning", "roaming profile gpo samba 4", "ipv6 preferred address", and "tomcat log"), so I don't even want to think what getting a debugging build is going to involve. @mqudsi, any thoughts?

**Comment by andrew-schulman** [comment]
FWIW the same problem happens in the 32- and 64-bit builds.

**Comment by zanchey** [comment]
Having thought about this some more, I'm not going to block 3.0.0 for this. Cygwin users will be no worse off than they are today ("use 2.7.1") and it will get a new release to other platforms' users sooner. We could look at a 3.0.1 if we can get it fixed well before 3.1. (sorry @andrew-schulman - not much of a Christmas present!)

**Comment by andrew-schulman** [comment]
S'ok. I don't disagree. When I can find some time, I'll do some debugging of my own, or ask on the Cygwin list.

**Comment by CodeMonkeyKevin** [comment]
Having the same issue on MacOS 10.13.

**Comment by zanchey** [comment]
@CodeMonkeyKevin, I'm surprised to hear that as we did lots more testing on macOS than Cygwin - I'm wondering if it's definitely the same issue. Are you able to run fish with `fish -d5` and confirm that it displays the same output as above?

**Comment by CodeMonkeyKevin** [comment]
@zanchey looks like it was a different issue. An issue with VCS ohm plugin.

**Comment by amzyang** [comment]
The same issue here. When cd into a git repo directory,  it blocks.
```
<4> fish: select_try: no fds returned valid data within the timeout
<5> fish: Skipping wait on incomplete job 6 (__fish_git_prompt_show_upstream ...)
<5> fish: Skipping wait on incomplete job 3 (__fish_git_prompt ...)
<5> fish: Skipping wait on incomplete job 1 (fish_prompt ...)
<4> fish: select_try on fd 9
```

Reproduce:

 `set __fish_git_prompt_showupstream 'auto'` or some other value.

```

    command git config -z --get-regexp '^(svn-remote\..*\.url|bash\.showupstream)$' 2>/dev/null | while read -lz key value
        switch $key
            case bash.showupstream
                set show_upstream $value
                test -n "$show_upstream"
                or return
            case svn-remote.'*'.url
                set svn_remote $svn_remote $value
                # Avoid adding \| to the beginning to avoid needing #?? later
                if test -n "$svn_url_pattern"
                    set svn_url_pattern $svn_url_pattern"|$value"
                else
                    set svn_url_pattern $value
                end
                set upstream svn+git # default upstream is SVN if available, else git

                # Save the config key (without .url) for later use
                set -l remote_prefix (string replace -r '\.url$' '' -- $key)
                set svn_prefix $svn_prefix $remote_prefix
        end
    end
```

**Comment by faho** [comment]
@amzyang: Which OS is this on?

**Comment by faho** [comment]
Simplifying a bit:

```
fish -c 'sleep 10 &; disown; set -g __fish_git_prompt_showupstream auto; __fish_git_prompt'
```

triggers it inside a git repo.

I need to try to simplify the git-prompt bit still.

Anyway, it seems that the presence of a disowned job can mess this up.

Enabling "full" job control seems to fix it.

**Comment by faho** [comment]
@amzyang: Your issue should be fixed in 4a3ac6e. 

The underlying problem was that we `disown`d a process via its pgid, which was one of those special values that caused us to reap the wrong process, which caused us to keep waiting for a process that was long gone.

@zanchey, @andrew-schulman: It'd be nice if either of you could check that commit to see if it's the same issue. I don't _think_ so, but you never know.

**Comment by mqudsi** [comment]
@zanchey is there a Cygwin VM image or EC2 AMI I could test this under? I have extremely painful memories about how long it takes to get Cygwin installed and how uncontained a Cygwin installation is.

I changed the behavior of how waiting within the context of a function works in 259cf02aacf3de83ccd9e12e434c65de5ec7ccd4, and although it may have some impact on this issue I am doubtful that it would make a real difference as we always keep enumerating jobs/processes and calling `waitpid(2)`, even after a job is reaped (just without suspending CPU if fg jobs are present).

**Comment by zanchey** [comment]
Still failing with 3.0.0-54-ga615151d on Cygwin. I'll see what I can do about getting you access to a VM, although getting Cygwin installed is fairly straightforward these days.

**Comment by mqudsi** [comment]
I'll try that the next chance I get to hack away at fish.

**Comment by amzyang** [comment]
> @amzyang: Your issue should be fixed in 4a3ac6e.

Thanks. The master branch works for me now.

**Comment by mqudsi** [comment]
@zanchey @andrew-schulman I just commited d1913f0df0e5d05fa553b6d482f1cb71261f22ba to `master` which gets fish running again. Make sure to build via cmake.

It seems to have been caused by some bugs in Cygwin's process management, specifically with regards to job reaping, that we did not previously run into with the past naive job control implementation.

The low-level tests still fail but I have to run and I did not get a chance to see why (none of them reported an error, for what it's worth).

**Comment by orthoxerox** [comment]
FYI, it hangs under MSYS2 as well, and setting it up for a debug build is much easier than Cygwin.

**Comment by orthoxerox** [comment]
MSYS2 uses [the following commands](https://github.com/Alexpux/MSYS2-packages/blob/master/fish/PKGBUILD) to build its version of fish, so I doubt a cmake fix will work as-is:

```
patch -p1 -i ${srcdir}/01-msysize.patch
autoreconf -ivf
./configure --prefix=/usr --with-doxygen --sysconfdir=/etc
make
make DESTDIR="${pkgdir}/" install
```

I hard-defined CYGWIN in `common.h`, and make install finished with a bunch of fork errors I failed to copy, but fish itself seems to work fine.

**Comment by andrew-schulman** [comment]
This problem is fixed for me now, building from master. Thanks!

Installation to a different destination directory doesn't work right, but that's a separate issue.

**Comment by ridiculousfish** [comment]
Reopening pending a pick/fix for 3.0.1

**Comment by zanchey** [comment]
The fix should probably just use `#ifdef __CYGWIN__` rather than adding to the CMake configuration checks.

**Comment by faho** [comment]
I have cherry-picked 4a3ac6e91ec415bfe536f722a71903d9a67bb10d, which is a slightly different issue.

I'm assuming we can just simplify the CYGWIN detection of d1913f0 for the actual issue?

**Comment by faho** [comment]
@mqudsi: I get a conflict if I try to cherry-pick these, and I'm not sure which part to take.

So should I simply add a

```c++
#ifdef __CYGWIN__
wait_by_process = true
#endif 
```

(sorry for the trailing space, github refuses to not autocomplete issue numbers after a #)

or would you prefer to resolve it?

**Comment by mqudsi** [comment]
No problem. Cherry picked for fish 3.0.1 as 91ac0f1b188d79f8b0d71fd2d50d967ecdd5e3bc and a5ef1e395e6016a145dbbdd72e175a50a3a2004f.
